crawler.rb

```diff
@@ -7,55 +7,55 @@
 require_relative './url_fetcher'
 
 class Crawler
-  attr_accessor :url, :href, :db
-  def initialize(db)
-    @database = db 
+  attr_accessor :url, :href, :connection
+  def initialize(connection)
+    @connection = connection 
   end
 
   def done_crawling?
-    @database.execute( "SELECT count(*) FROM URLS_to_crawl WHERE state = 'uncrawled'" )[0][0] == 0
+    @connection.execute("SELECT count(*) FROM URLS_to_crawl WHERE state = 'uncrawled'")[0][0] == 0
   end
 
   def crawl_next_url
     begin
-      url = @database.execute( "SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1" )[0][0]
+      url = @connection.execute("SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1")[0][0]
       content = UrlFetcher.fetch(url)
       document = Document.new(url, content)
-      write_documents_to_db(document)
-      document.domain_hrefs.each do |href|
-        write_urls_to_db(href)
-      end
+      get_url_and_write_document(document)
+      @connection.transaction
```

Comment:
I think this would be better as a block call. What happens if you iterator below raises an exception mid-transaction?